### PR TITLE
1351 character issues

### PR DESCRIPTION
**Pull request by HalcyonJAC**
## What's included?
See parent ticket for details

## Who should test?
✅ Product owner
✅ Developers
✅ UTG

## How to test?
- Choose an exercise (i.e. Warren Legal)
- Click the Character Issues option from the Reports menu
- If there is no data on-screen, click the Refresh button (to generate the character issue data)
- Use the dropdown to filter the data on-screen if you like
- Click the Export Data button to download the data in XLS format

## Risk - how likely is this to impact other areas?
🟠 Medium risk - this does change code that is shared with other areas

PREVIEW:DEVELOP

### CODE REVIEW

**Comment by joy-ade** [comment]
@HalcyonJAC  I am sending this out to UTG

**Comment by mbrookeswebdev** [comment]
I can see some code related to assessors ticket in this PR, should it be here(?)

With regards to the Character Issues:

It probably applies to quite a few places in our app but where there is no data, the page displays nothing. Maybe we should display a message to that effect.

I've applied to an exercise JAC00298, which uses character info v2 [as this candidate](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/KOpW1w2nCyVe83rVFlz6/applications/applied/application/x7zPCnHFbcF9rqdueU0q). 

- When I've checked out [Character Issues page](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/KOpW1w2nCyVe83rVFlz6/reports/character) it displays blank where the issues should be listed, but the number of issues in the downloaded spreadsheet is correct.

I've applied to the old exercise JAC0053 as [the same candidate](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/application/RCUWUL01BKbCT3DCbF0z#full)

- On Character Issues page when I've selected All applications in the drop-down and click Next on the bottom of the page it takes me to the top of the same page instead of the next page.

- Character Issues displays blank where the issues should be listed. The downloaded report contains mixed info from both applications.

I've also tested on another candidate with less data ([old exercise](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/applications/applied/application/KtPwJAreAOaBZWP1Pcp9) and [new exercise](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/KOpW1w2nCyVe83rVFlz6/applications/applied/application/7VAfk16N52pKnaJ4mtby)), it seems downloaded data is mixed from both applications.

A couple of screenshots:
screenshot1
Screenshot2

**Review by warrensearle — CHANGES_REQUESTED** [review]
✅ The on screen report appears to be working well
✅ The export data download is working
✅ The code looks fine though see issue below

The following needs to be addressed:

1. This PR includes changes from the IAs validation work, so it's not possible to approve this without approving both changes. Are you able to disentangle the two pieces of work? If not then we should not merge this until IAs validation has been approved (and merged).

2. The downloaded file doesn't include the issues and events displayed on screen. In fact it looked like the wrong download.

3. The Stages and Status dropdowns work well but I wonder if the layout could be tidied. On desktop right now they appear one above the other which pushes the report content quite far down the screen. 
image
Having the dropdowns side-by-side might help.

**Comment by mbrookeswebdev** [comment]
@HalcyonJAC I've done a quick check. The issues appear on the page, which is great.

When I click 'Next' [on this page](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/reports/character#) it still takes me to the top of the same page, not to the next page with candidates.

Also, the downloaded data still seems to be incorrect. For example [this candidate](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/KOpW1w2nCyVe83rVFlz6/application/7VAfk16N52pKnaJ4mtby#full) has 3 issues listed in the app, 5 issues on the Char Issues page and 6 in the download file.

**Comment by HalcyonJAC** [comment]
> 3\. The Stages and Status dropdowns work well but I wonder if the layout could be tidied. On desktop right now they appear one above the other which pushes the report content quite far down the screen.

I have made a styling change so that the dropdowns are now side-by-side. Note: I tried doing this using the gov-uk classes but couldn't get it to work correctly that way, so used a small amount of CSS instead.

**Comment by HalcyonJAC** [comment]
> Also, the downloaded data still seems to be incorrect. For example [this candidate](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/KOpW1w2nCyVe83rVFlz6/application/7VAfk16N52pKnaJ4mtby#full) has 3 issues listed in the app, 5 issues on the Char Issues page and 6 in the download file.

I checked this out and the `applications` document has both `characterInformation` and `characterInformationV2` stored against it (see below), which is what is causing the inconsistancy. In production this (bad data) shouldn't happen.

image

**Comment by warrensearle** [comment]
>I checked this out and the applications document has both characterInformation and characterInformationV2 stored against it (see below), which is what is causing the inconsistancy. In production this (bad data) shouldn't happen.
@HalcyonJAC I like your optimism however thinking perhaps we could code against this anyway. If `characterInformationV2` exists then ignore `characterInformation`. Being more pessimistic I think we _could_ have applications with both fields ☹️

**Comment by HalcyonJAC** [comment]
> When I click 'Next' [on this page](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/reports/character#) it still takes me to the top of the same page, not to the next page with candidates.

I don't know how to fix this!

**Comment by mbrookeswebdev** [comment]
> I checked this out and the `applications` document has both `characterInformation` and `characterInformationV2` stored against it (see below), which is what is causing the inconsistancy. In production this (bad data) shouldn't happen.

If I understand correctly, the data in the report should contain info from [the Character Information section of the application](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/application/KtPwJAreAOaBZWP1Pcp9#full). Couldn't we do it like this - if `(this.exercise._applicationVersion && this.exercise._applicationVersion === 2) then use characterInformationV2, if not then use characterInformation`. [This is how the correct info displayed in Full details tab.](https://github.com/jac-uk/admin/blob/main/src/views/Exercises/Applications/Application.vue)(line 317 onwards).

> When I click 'Next' on this page it still takes me to the top of the same page, not to the next page with candidates.
I don't know how to fix this!

With regards to the table, it is a quite complex component... [Maybe my WIP branch would help](https://github.com/jac-uk/admin/compare/feature/1327-enable-request-of-character-check-form-completion#diff-6f061963ddb50672ea6134049e69c0e2111a0071e2c90f16045bfc85a3e27ab3), check out how I've used the table in `src/views/Exercises/Tasks/CharacterChecks.vue` for my consent form prototype.

**Comment by HalcyonJAC** [comment]
> I can see some code related to assessors ticket in this PR, should it be here(?)

I must have pushed the assessor changes onto this branch by mistake. I have removed these changes from this branch now.

**Comment by HalcyonJAC** [comment]
> If I understand correctly, the data in the report should contain info from [the Character Information section of the application](https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app/exercises/wdpALbyICL7ZxxN5AQt8/application/KtPwJAreAOaBZWP1Pcp9#full). Couldn't we do it like this - if `(this.exercise._applicationVersion && this.exercise._applicationVersion === 2) then use characterInformationV2, if not then use characterInformation`. [This is how the correct info displayed in Full details tab.](https://github.com/jac-uk/admin/blob/main/src/views/Exercises/Applications/Application.vue)(line 317 onwards).

The way I have done it is by altering the `flagApplicationIssues` function in the Digital Platform (after a conversation with Warren) so that the issue details end up being part of the `applicationRecords.characterIssues` array, whether the source was character info v1 or v2. The reason it isn't working right for the application you picked is that is happens to have character info v1 and v2 on it, which wouldn't usually happen.

**Review by warrensearle — APPROVED** [review]

File: src/views/Exercises/Show/Reports/CharacterIssues.vue [hunk 1]
```diff
@@ -27,17 +27,77 @@
       </button>
     </div>
 
+    <div class="govuk-grid-column-full">
```
Comment (tomlovesgithub) [inline src/views/Exercises/Show/Reports/CharacterIssues.vue:30]: This should sort the styling and remove the need for custom CSS
```
...
    <div class="govuk-grid-column-full">
      <div class="govuk-button-group">
        <Select
          id="exercise-stage"
          v-model="exerciseStage"
          class="govuk-!-margin-right-2"
        >
          <option 
...
```

**Review by tomlovesgithub — APPROVED** [review]
Look good to me, y only suggestions would be the CSS changes mentioned above and perhaps renaming the download, which could perhaps be expanded into another ticket so the filename carries filtering information.

**Comment by HalcyonJAC** [comment]
Hi 

The preview URL for this is https://jac-admin-develop--pr1352-feature-1351-charact-xkdr9f3a.web.app

...and it does not expire until Sun, 27 Jun 2021

**Comment by joy-ade** [comment]
This needs changes post qualification experience is missing from the download

File: src/views/Exercises/Show/Reports/CharacterIssues.vue (hunk 1 above)
Comment (HalcyonJAC) [reply src/views/Exercises/Show/Reports/CharacterIssues.vue:30]: Thanks for @tomlovesgithub, worked like a charm :)

**Comment by HalcyonJAC** [comment]
> This needs changes post qualification experience is missing from the download

I have fixed this now.

**Review by joy-ade — APPROVED** [review]
Approved by product team and UTG

**Pull request merged**